stdlib/Profile/src/Allocs.jl

```diff
@@ -0,0 +1,170 @@
+end
+
+function _prof_expr(expr, opts)
+    quote
+        $start(; $(esc(opts)))
+        local res = $(esc(expr))
+        $stop()
+        res
+    end
+end
+
+function start(; sample_rate::Number)
+    ccall(:jl_start_alloc_profile, Cvoid, (Cdouble,), Float64(sample_rate))
+end
+
+function stop()
+    ccall(:jl_stop_alloc_profile, Cvoid, ())
+end
+
+function clear()
+    ccall(:jl_free_alloc_profile, Cvoid, ())
+end
+
+function fetch()
+    raw_results = ccall(:jl_fetch_alloc_profile, RawAllocResults, ())
+    decoded_results = decode(raw_results)
+    @warn("This allocation profiler currently misses some allocations. " *
+    "For more info see https://github.com/JuliaLang/julia/issues/43688")
+    return decoded_results
+end
+
+# decoded results
+
+struct Alloc
+    type::Any
+    stacktrace::StackTrace
+    size::Int
+end
+
+struct AllocResults
+    allocs::Vector{Alloc}
+end
+
+# Without this, the Alloc's stacktrace prints for lines and lines and lines..
+function Base.show(io::IO, a::Alloc)
+    stacktrace_sample = length(a.stacktrace) >= 1 ? "$(a.stacktrace[1]), ..." : ""
+    print(io, "$Alloc($(a.type), $StackFrame[$stacktrace_sample], $(a.size))")
+end
+
+const BacktraceCache = Dict{BTElement,Vector{StackFrame}}
+
+# copied from julia_internal.h
+const JL_BUFF_TAG = UInt(0x4eadc000)
+
+struct CorruptType end
+struct BufferType end
+
+function load_type(ptr::Ptr{Type})
+    if UInt(ptr) < UInt(4096)
+        return CorruptType
+    elseif UInt(ptr) == JL_BUFF_TAG
+        return BufferType
+    end
+    return unsafe_pointer_to_objref(ptr)
+end
+
+function decode_alloc(cache::BacktraceCache, raw_alloc::RawAlloc)::Alloc
+    Alloc(
+        load_type(raw_alloc.type),
+        stacktrace_memoized(cache, load_backtrace(raw_alloc.backtrace)),
+        UInt(raw_alloc.size)
+    )
+end
+
+function decode(raw_results::RawAllocResults)::AllocResults
+    cache = BacktraceCache()
+    allocs = [
+        decode_alloc(cache, unsafe_load(raw_results.allocs, i))
+        for i in 1:raw_results.num_allocs
+    ]
+    return AllocResults(allocs)
+end
+
+function load_backtrace(trace::RawBacktrace)::Vector{BTElement}
+    out = Vector{BTElement}()
+    for i in 1:trace.size
+        push!(out, unsafe_load(trace.data, i))
+    end
+
+    return out
+end
+
+function stacktrace_memoized(
+    cache::BacktraceCache,
+    trace::Vector{BTElement},
+    c_funcs::Bool=true
+)::StackTrace
+    stack = StackTrace()
+    for ip in trace
+        frames = get(cache, ip) do
+            res = lookup(ip)
+            cache[ip] = res
+            return res
+        end
+        for frame in frames
+            # Skip frames that come from C calls.
+            if c_funcs || !frame.from_c
+                push!(stack, frame)
+            end
+        end
+    end
+    return stack
+end
+
+# Precompile once for the package cache,
+@assert precompile(start, ())
+@assert precompile(stop, ())
+@assert precompile(fetch, ())
+
```

Comment:
So @vilterp and I noticed something strange about the performance, though, that we'd like to understand and fix. I'm not sure if we can figure it out on our own.. @vchuravy or @vtjnash if you could lend your aid that would be great!

Locally, we observed that enabling the Allocations Profiler with `sample_rate=1` (it defaults to 0.0001 to avoid this overhead) makes this example about 10x-20x slower:
```julia
julia> @time for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.039658 seconds (49.11 k allocations: 3.107 MiB, 9.16% compilation time)

julia> @time Allocs.@profile sample_rate=1 for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.549963 seconds (49.11 k allocations: 3.107 MiB, 65.56% compilation time)
```
We were trying to understand where this change comes from, and so we took some CPU profiles of the above and compared them in PProf:
```julia
julia> @time @profile for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.047540 seconds (49.11 k allocations: 3.107 MiB, 11.30% compilation time)
```
```julia
julia> @time @profile Profile.Allocs.@profile sample_rate=1 for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.609904 seconds (49.98 k allocations: 3.160 MiB, 65.49% compilation time)
```

With the allocations profiler, about 45% of the time is spent in recording the stacktraces (in `jl_unw_stepn`), and (from what i understand) there's really nothing we can do about that. This is the same image we put in the PR description:
<img width="300" alt="148444290-2ed5a7b2-5b26-4dde-827e-35c22219cfe2" src="https://user-images.githubusercontent.com/1582097/148577086-f86b96f0-c12c-42fd-8f00-be1b0f2c9cb9.png">

**However:** We _also_ see that another 45% of the time is spent in these top-level `<unknown>` frames, which aren't there when we run without the allocations profiler. We don't know what these are, nor where they're coming from. We _think_ we remember learning that the unknown frames are likely to be frames spent in the Julia Interpreter instead of compiled frames? Do you have any idea what this could be interpreting, and why?

Here's the before/after comparison - first **without** Allocations Profiling, and then **with it enabled**:
<img width="1196" alt="Screen Shot 2022-01-06 at 3 32 56 PM" src="https://user-images.githubusercontent.com/1582097/148577388-f7e8fe47-fcfa-418f-8de2-73ce67734784.png">

<img width="1223" alt="Screen Shot 2022-01-06 at 3 33 18 PM" src="https://user-images.githubusercontent.com/1582097/148577496-544b8782-b51c-4ca6-ab67-98f1629dca23.png">

Finally, for clarity, here I've highlighted in purple our `record_allocated_value()` frame, which is our entry point to stack unwinding. You can see that this accounts for 45% of the profile after the change (or 90% of the time without that extra <unknown> thing).

------------------------------


Can you help us figure out what this is? It's not _urgent_, because with a smaller sample rate like even 0.01 or 0.001, the overhead is almost entirely absent. So I think we could merge this without identifying and fixing it, but it would be nice to address. I think there's potentially a 2x reduction in the overhead that's available here. 👍 Thanks in advance!!

Reply:
EDIT: Here's the annotated photo, which illustrates our question:

<img width="1202" alt="Screen Shot 2022-01-06 at 3 33 36 PM" src="https://user-images.githubusercontent.com/1582097/148588957-00c87788-6751-4797-b664-bc70d8a33884.png">

Reply:
No spontaneous theories, but I would use `perf` to get a second opinion :)

Reply:
yeah, that's a good idea, thanks!